Add multi-process, multi-thread and PSA storage partition support.

simonqhughes opened the pull request:
The following provides more information on this commit:
- Add getuid() user ID to file object basename. The user ID is used as the
  partition identifier.
- Add process and thread IDs to temporary filename basename. This makes accesses
  to the temporary files process and thread safe, as there is only 1 temporary
  file in existence at a time.
- Make mktemp_filename() temporary filenames process and thread safe.
- Add psa_test_case_201.sh for multi-process multi-thread testing.
- Add additional test cases to psa-storage-example-app for multi-process
  multi-thread testing.

A jenkins job for testing can be found here:

http://jenkins.mbed-linux.arm.com/view/simon/job/sdh_zeus_dev_test_1/4/console

The following shows the trace of test application binary running on the imx8:

```
root@mbed-linux-os-678:~# psa-storage-example-app 


PSA Storage Tests Derived from mbed-crypto Tests
================================================

Set/get/remove 0 bytes ............................................ PASS
Set/get/remove 42 bytes ........................................... PASS
Set/get/remove 1000 bytes ......................................... PASS
Set/get/remove with flags ......................................... PASS
Overwrite 0 -> 3 .................................................. PASS
Overwrite 3 -> 0 .................................................. PASS
Overwrite 3 -> 3 .................................................. PASS
Overwrite 3 -> 18 ................................................. PASS
Overwrite 18 -> 3 ................................................. PASS
Multiple files .................................................... PASS
Non-existent file ................................................. PASS
Removed file ...................................................... PASS
Get 0 bytes of 10 at 10 ........................................... PASS
Get 1 byte of 10 at 9 ............................................. PASS
Get 0 bytes of 10 at 0 ............................................ PASS
Get 1 byte of 10 at 0 ............................................. PASS
Get 2 bytes of 10 at 1 ............................................ PASS
Get 1 byte of 10 at 10: out of range .............................. PASS
Get 1 byte of 10 at 11: out of range .............................. PASS
Get 0 bytes of 10 at 11: out of range ............................. PASS
Get -1 byte of 10 at 10: out of range ............................. PASS
Get 1 byte of 10 at -1: out of range .............................. PASS
Flags test ........................................................ PASS
Set/get/remove 0 bytes ............................................ PASS
Set/get/remove 42 bytes ........................................... PASS
Set/get/remove 1000 bytes ......................................... PASS
Set/get/remove with flags ......................................... PASS
Overwrite 0 -> 3 .................................................. PASS
Overwrite 3 -> 0 .................................................. PASS
Overwrite 3 -> 3 .................................................. PASS
Overwrite 3 -> 18 ................................................. PASS
Overwrite 18 -> 3 ................................................. PASS
Multiple files .................................................... PASS
Non-existent file ................................................. PASS

----------------------------------------------------------------------------

PASSED (34 / 34 tests (0 skipped))
Execute_tests                  PASS

Robustness Against Power Failure Test Cases
===========================================

TC001: Rcvr .dat (0 .dat, 2 .bak, F_WRITE_ONCE unset)                  PASS
TC002: Rcvr .dat from 0 dat, 1 .bak, F_WRITE_ONCE unset                PASS
TC051: Rcvr .dat (1 .dat (new), 2 .bak, F_WRITE_ONCE unset)            PASS
TC052: Rcvr .dat (1 .dat (old), 2 .bak, F_WRITE_ONCE unset)            PASS
TC053: Rcvr .dat (1 .dat(new, mismatched), 1 .bak, F_WRITE_ONCE unset) PASS
TC054: Rcvr .dat (1 .dat (old, mismatched), 1 .bak, F_WRITE_ONCE unset PASS
TC055: Rcvr .bak (1 .dat, 0 .bak, F_WRITE_ONCE unset)                  PASS
TC101: Rcvr .dat (0 .dat, 2 .bak, F_WRITE_ONCE set)                    PASS
TC102: Rcvr .dat (0 dat, 1 .bak, F_WRITE_ONCE set)                     PASS
TC151: Rcvr .dat (1 .dat (new), 2 .bak, F_WRITE_ONCE set)              PASS
TC152: Rcvr .dat (1 .dat (old), 2 .bak, F_WRITE_ONCE set)              PASS
TC153: Rcvr .dat (1 .dat (new, mismatched), 1 .bak, F_WRITE_ONCE set)  PASS
TC154: Rcvr .dat (1 .dat (old, mismatched), 1 .bak, F_WRITE_ONCE set)  PASS
TC155: Rcvr .bak (1 .dat, 0 .bak, F_WRITE_ONCE set)                    PASS
TC201: Multi-process support (F_WRITE_ONCE unset)                      PASS
TC203: Multi-process-Multi-thread support (F_WRITE_ONCE unset)         PASS
```

simonqhughes committed:
Add multi-process, multi-thread and PSA storage partition support.

The following provides more information on this commit:
- Add getuid() user ID to file object basename. The user ID is used as the
  partition identifier.
- Add process and thread IDs to temporary filename basename. This makes accesses
  to the temporary files process and thread safe, as there is only 1 temporary
  file in existence at a time.
- Make mktemp_filename() temporary filenames process and thread safe.
- Add psa_test_case_201.sh for multi-process multi-thread testing.
- Add additional test cases to psa-storage-example-app for multi-process
  multi-thread testing.

simonqhughes committed:
SQ: Remove debug options from lib/Makefile .c.o rule. Cleanup lib/common_storage.c

simonqhughes committed:
SQ: Fix common_storage.c warnings.

diego-sueiro commented on test/psa_test_case_201.sh:
```diff
@@ -0,0 +1,91 @@
+    ${PSA_CMD_ID} ${user} > /dev/null 2>&1
+    if [ "$?" -eq "1" ]; then
+        # user doesnt exist
+        sudo useradd ${user}
```
Have you tested this in one of our targets?
Our images don't have `bash` and `sudo` for example. How this is script ran on imx8?

simonqhughes replied:
The script is for testing on x86. Its useful to check the required properties of the implementation.

diego-sueiro replied:
Right. Can you please add a comment in the script about this.
Also, the other tests from `test/Makefile` are supposed to run only on x86?

Is there any plan to have these tests running on our supported platforms?

diego-sueiro replied:
I also think that the instructions that you put on the top of the `test/Makefile` should go into a `test/README.md` file.

simonqhughes replied:
>> Have you tested this in one of our targets?

Yes, on imx8. The test output shown in the PR summary shows the test results collected on imx8.

diego-sueiro replied:
But you said earlier that this script is for testing on x86.

simonqhughes replied:
>> But you said earlier that this script is for testing on x86.

The documentation in the following HLD (https://confluence.arm.com/display/mbedlinux/PSA+Storage+Software+HLD) sections help answer this question.
- Test Case 201: Multi-process support test
- Test Case 203: Multi-process-multi-thread support test

If these sections are insufficient then lets discuss together the details.

simonqhughes replied:
>> Right. Can you please add a comment in the script about this.

Yes, I'll do that.

>> Also, the other tests from test/Makefile are supposed to run only on x86?

The short answer is no. Testing can be performed on x86 and on-target. Here are some more details to explain this:
- The project builds a test application (`psa-storage-example-app`). This is built to include the test cases.
- `psa-storage-example-app` can be run on target. The test results output are shown above at the description for this PR.
- `psa-storage-example-app` can be also be built for x86, and run multiple times in parallel with the `psa_test_case_201.sh` bash script.
- The project also builds x86 test binaries derived from the reference PSA test framework project psa-arch-tests. The binaries generated are`psa-arch-tests-ps`, `psa-arch-tests-its`, `psa-arch-tests-crypto`. These are for testing the PSA PS,  ITS and Crypto APIs respectively. 

Hope this helps.

simonqhughes replied:
>> Right. Can you please add a comment in the script about this.

The above feedback comment has been addressed with commit https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/00cf943e90fdff4e9f6287f8ff5368a5cbd8d5af.

simonqhughes replied:
>> I also think that the instructions that you put on the top of the test/Makefile should go into a test/README.md file.

The above feedback comment has been addressed with commit https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/ff7c3205cf389cca773aee433e7fc1146be4e7b8

simonqhughes committed:
SQ: Fix Makefiles and code to stringify PSA_STORAGE_FILE_C_STORAGE_PREFIX.

simonqhughes committed:
SQ: Remove -Werror from app/Makefile and lib/Makefile.

diego-sueiro commented:
I think we should bump the version for this implementation and also add the version information in the Makefile alongside with a git tag. The `psa-trusted-storage-linux` recipe name will also need to be updated.

simonqhughes committed:
SQ: Update test/psa_test_case_201.sh comments.

The following provides more information on this commit:
- The comments at the start of the script have been updated to document
  that the script is currently used for x86 testing.

simonqhughes committed:
SQ: Add test/README.md and document the use of test/Makefile.

The following provides more information on this commit:
- The comments at the start of the test/Makefile have been moved into a new
  test/README.md file.

rwalton-arm commented on app/main.c:
```diff
@@ -1672,14 +1657,15 @@ int main( int argc, const char *argv[] )
             NULL
             };
 
     const char* tc_description[] = {
```
This array should be `static const char*` if we're not mutating it later on in the code.

simonqhughes replied:
Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/3814d49ae609c6052baf68494f6585eecf4b1324.

simonqhughes reviewed (commented):
>> Can you please explain me the tag layout structure used in this repo?

Jezz settled on the following policy. Tags are created for the downstream consumers e.g. the PDMC team (Jenia Kogan et al.) for their releases. There has only been 1 tag created for them: `ptsl-0.0.1-psa-storage-api-1.0.0`. 
- `pstl` is the acronym from the name of this repo (PSA Trusted Storage Linux).
- The first version tuple is specified by us. When we tag, i suggest this version tuple should match whats in `<topdir>/Makefile`.
- `psa-storage-api` qualifies the second version tuple.
- The second version tuple is the version of the PSA Storage API supported by the implementation. 

>> I think we should bump the version for this implementation

OK, but I think this needs some discussion. Consider:
- If we're following the above policy there is no need to bump the version until the next "release" is made.
- The tuple `1.0.0` hasn't been used in a release yet.
- The `ptsl-0.0.1-psa-storage-api-1.0.0` tag sha is the one that appears in the `meta-mbl psa-trusted-storage-linux_1.0.0.bb` recipe. The Makefile version information is not present in `ptsl-0.0.1-psa-storage-api-1.0.0::<topdir>/Makefile`. Perhaps this inconsistency is best corrected by renaming `psa-trusted-storage-linux_1.0.0.bb` to `psa-trusted-storage-linux_0.0.1.bb`? When the recipe sha is updated to point to tuple x.y.z then we can rename the recipe to `psa-trusted-storage-linux_x.y.z.bb`.

>> and also move the version information in the Makefile to the top of the file

Do you mean move <topdir>/Makefile lines 23-25 to lines 16-18?

>>  alongside with a git tag

Do you mean adding a comment which specifies a git tag thats created after the PR is merged? or something else?

diego-sueiro commented:
> > > Can you please explain me the tag layout structure used in this repo?
> 
> Jezz settled on the following policy. Tags are created for the downstream consumers e.g. the PDMC team (Jenia Kogan et al.) for their releases. There has only been 1 tag created for them: `ptsl-0.0.1-psa-storage-api-1.0.0`.
> 
> * `pstl` is the acronym from the name of this repo (PSA Trusted Storage Linux).
> * The first version tuple is specified by us. When we tag, i suggest this version tuple should match whats in `<topdir>/Makefile`.
> * `psa-storage-api` qualifies the second version tuple.
> * The second version tuple is the version of the PSA Storage API supported by the implementation.
> 

Understood. Thanks. And yes, the version information in the project Makefile must match the first version tuple.

> > > I think we should bump the version for this implementation
> 
> OK, but I think this needs some discussion. Consider:
> 
> * If we're following the above policy there is no need to bump the version until the next "release" is made.

As I understand, we are delivering a new features for PSA Storage supporting multi-process/multi-threads. To _celebrate_ this I think we have to bump at least the `rel` version.

> * The tuple `1.0.0` hasn't been used in a release yet.
> * The `ptsl-0.0.1-psa-storage-api-1.0.0` tag sha is the one that appears in the `meta-mbl psa-trusted-storage-linux_1.0.0.bb` recipe. The Makefile version information is not present in `ptsl-0.0.1-psa-storage-api-1.0.0::<topdir>/Makefile`. Perhaps this inconsistency is best corrected by renaming `psa-trusted-storage-linux_1.0.0.bb` to `psa-trusted-storage-linux_0.0.1.bb`? When the recipe sha is updated to point to tuple x.y.z then we can rename the recipe to `psa-trusted-storage-linux_x.y.z.bb`.
This needs to be fixed in the recipe name.

> 
> > > and also move the version information in the Makefile to the top of the file
> 
> Do you mean move /Makefile lines 23-25 to lines 16-18?
> 

Yes, and we should not use the `?=` assignment for the versioning setting.

> > > alongside with a git tag
> 
> Do you mean adding a comment which specifies a git tag thats created after the PR is merged? or something else?

No. What I meant was: if we bump the version in the make file we need to create a new tag (following the proposed tag layout policy).

simonqhughes committed:
SQ: Update <topdir>/Makefile version symbols from 1.0.0 to 1.0.1.

The following provides more information on this commit:
- The psa_trusted_storage_linux version information has been updated from
  1.0.0. to 1.0.1.

simonqhughes committed:
SQ: Make  <topdir>/app/main.c test case description array static.

simonqhughes commented:
>> To celebrate this I think we have to bump at least the rel version.

>> Yes, and we should not use the ?= assignment for the versioning setting.

The `<topdir>/Makefile` version information has been fixed in commit https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/e06a7623ac5fba923cabf6c3e4662a6f287d15ea

diego-sueiro commented:
> > > To celebrate this I think we have to bump at least the rel version.
> 
> > > Yes, and we should not use the ?= assignment for the versioning setting.
> 
> The `<topdir>/Makefile` version information has been fixed in commit [e06a762](https://github.com/ARMmbed/psa_trusted_storage_linux/commit/e06a7623ac5fba923cabf6c3e4662a6f287d15ea)

Shouldn't be 0.0.2? Why the major version is 1? I think we agreed that the version in the Makefile must match with the first version tuple of the git tag `ptsl-x.y.z-psa-storage-api-1.0.0`. So we need to have on the Makefile:
```
major = x
minor = y
rel  = z
```

simonqhughes committed:
SQ: Update <topdir>/Makefile version symbols from 1.0.1 to 0.0.2.

simonqhughes commented:
>> Shouldn't be 0.0.2? 

Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/9ad6f0959b6adbd28614da09bed4a37048fcf12f

simonqhughes committed:
SQ: Update <topdir>/README.md to document API support and application multi-thread synchronisation.

diego-sueiro commented on app/Makefile:
```diff
@@ -22,9 +22,10 @@ RM = rm -f
-LOCAL_CFLAGS ?= -Wall -W -Wdeclaration-after-statement -I../inc
-LOCAL_LDFLAGS = ../lib/libpsastorage.a
+#CFLAGS ?= -g3 -gstabs+
+CFLAGS ?=
```
These lines can be removed.

simonqhughes replied:
Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/5206b43d1546c452d99c5c61624458c94670bcb5

diego-sueiro commented on lib/Makefile:
```diff
@@ -18,9 +18,10 @@ INSTALL = install
 
-CFLAGS ?= -O2 -g
+#CFLAGS ?= -g3 -gstabs+
+CFLAGS ?=
```
These lines can be removed.

simonqhughes replied:
Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/5206b43d1546c452d99c5c61624458c94670bcb5

diego-sueiro commented on lib/Makefile:
```diff
@@ -18,9 +18,10 @@ INSTALL = install
+CFLAGS ?=
 LOCAL_CFLAGS ?= -fPIC -Wall -Wextra -I../inc
-LOCAL_LDFLAGS ?= -shared
+LOCAL_LDFLAGS ?= -shared -lpthread
```
`?=` shouldn't be used and the `-lpthread` should be passed when `PSA_STORAGE_TEST` is present. You can have something like this:
```
LOCAL_LDFLAGS = -shared
ifeq ($(PSA_STORAGE_TEST),1)
LOCAL_LDFLAGS += -lpthread 
CFLAGS += -DPSA_STORAGE_TEST
endif
```

Then the make command for the test can be something like:
```
make PSA_STORAGE_TEST=1
```

simonqhughes replied:
Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/5206b43d1546c452d99c5c61624458c94670bcb5

diego-sueiro commented on test/Makefile:
```diff
@@ -184,6 +157,10 @@ psa-storage-install: psa-storage
+.PHONY: test-psa-storage-app-mt
+test-psa-storage-app-mt: $(TEST_DIR) psa-storage
+	export LD_LIBRARY_PATH=$(LD_LIBRARY_PATH) && export PATH=$(PATH):$(TOPDIR)/$(PSA_STORAGE_DIR)/$(PREFIX)/bin && $(PSA_STORAGE_DIR)/test/$(PSA_STORAGE_MULTI_THREAD_TEST_SCRIPT)
+
```
In some lines bellow, for the `ws-create` and `ws-create-pinned` targets, why are you doing `repo init -u https://github.com/armmbed/mbl-manifest`?

These steps are very confusing. You are cloning the `psa_trusted_storage_linux` repo again. Why? If the repo is already cloned you don't need to clone it again, copy the manifest and sync. This is overcomplicated. You could have something link this:
```
repo init -u https://github.com/ARMmbed/psa_trusted_storage_linux -m test/psa-storage-test-manifest.xml
```

Also, please get rid of the ssh protocol URLs on this Makefile and on the manifests. All repos are public and we can use the https.

simonqhughes replied:
Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/2a19feccd83e088688d861e15faa7b26ae158821, which also adds psa-arch-tests patches.

diego-sueiro commented on test/README.md:
```diff
@@ -0,0 +1,49 @@
+ - psa_trusted_storage_linux test binaries (derived from the mbed-crypto
+   tests) e.g. `psa-storage-example-app`.
+
+(WARNING: Public users will be unable to use this makefile to generate
```
I don't think we have nay private repo.

simonqhughes replied:
Agree, will fix this.

simonqhughes replied:
Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/5206b43d1546c452d99c5c61624458c94670bcb5

diego-sueiro commented on test/README.md:
```diff
@@ -0,0 +1,49 @@
+The projects of interest are:
+ - https://github.com/ARMmbed/psa_trusted_storage_linux.
+ - https://github.com/ARMmbed/mbed-crypto.
+ - A private fork of https://github.com/ARM-software/psa-arch-tests. The private
```
The fork is not private.

simonqhughes replied:
Yes, I'll fix this.

simonqhughes replied:
Fixed in https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/5206b43d1546c452d99c5c61624458c94670bcb5

diego-sueiro commented on test/psa_test_case_201.sh:
```diff
@@ -0,0 +1,96 @@
+# !/bin/bash
```
```suggestion
#!/bin/bash
```

diego-sueiro replied:
Can you please run and fix the errors reported by the shellcheck in this file?

simonqhughes replied:
Will do.

simonqhughes replied:
`shellcheck` changes added in commit https://github.com/ARMmbed/psa_trusted_storage_linux/pull/17/commits/0641e0421c3edc5dbae6fc460a32c75f034d3929.

simonqhughes committed:
SQ: Update <topdir>/test/psa_test_case_201.sh with shellcheck fixes.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -29,29 +29,114 @@
 #include <stdlib.h>
 #include <inttypes.h>       /* for PRIu64 */
 #include <errno.h>          /* for errno */
+#include <sys/syscall.h>    /* for gittid() */
```
```suggestion
#include <sys/syscall.h>    /* for gettid() */
```

simonqhughes replied:
If this work is restarted, this change can be accepted.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -29,29 +29,114 @@
+/* PSA_CS_FILENAME_XUID_PATTERN is the base filename pattern used for RUID, PID and TID patterns. */
+#define PSA_CS_FILENAME_XUID_PATTERN "%08lx_"
+
+/* PSA_CS_FILENAME_XUID_PATTERN_LEN is the XUID pattern string length.
```
What is the length unit? Is it in bytes? Can you please specify it in the comment?

simonqhughes replied:
If this work is restarted, this change can be made.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -29,29 +29,114 @@
+/* PSA_CS_FILENAME_UID_PATTERN is the PSA Storage Unique ID (UID) pattern. */
+#define PSA_CS_FILENAME_UID_PATTERN "%08lx%08lx"
+
+/* PSA_CS_FILENAME_UID_PATTERN_LEN is the PSA_CS_FILENAME_UID_PATTERN string length. */
```
What is the length unit? Is it in bytes? Can you please specify it in the comment?

simonqhughes replied:
If this work is restarted, this change can be made.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -29,29 +29,114 @@
+ * object filename <RUID>_<UID>.dat. */
+#define PSA_CS_FILENAME_RUID_PATTERN PSA_CS_FILENAME_XUID_PATTERN
+
+/* PSA_CS_FILENAME_RUID_PATTERN_LEN is the PSA_CS_FILENAME_RUID_PATTERN string length. */
```
What is the length unit? Is it in bytes? Can you please specify it in the comment?

simonqhughes replied:
If this work is restarted, this change can be made.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -29,29 +29,114 @@
+ * filename <RUID>_<UID>.dat. */
+#define PSA_CS_FILENAME_PATTERN     PSA_CS_FILENAME_RUID_PATTERN PSA_CS_FILENAME_UID_PATTERN
+
+/* PSA_CS_FILENAME_PATTERN_LEN is the PSA_CS_FILENAME_PATTERN string length. */
```
What is the length unit? Is it in bytes? Can you please specify it in the comment?

simonqhughes replied:
If this work is restarted, this change can be made.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -29,29 +29,114 @@
+#define PSA_CS_LOCK_FILENAME_PATTERN                    PSA_CS_TMP_FILENAME_PATTERN
+#define PSA_CS_OFD_LOCK_FILENAME_PATTERN                PSA_CS_FILENAME_RUID_PATTERN
+
+/* PSA_CS_TMP FILENAME_PATTERN_LEN is the PSA_CS_FILENAME_PATTERN string length. */
```
What is the length unit? Is it in bytes? Can you please specify it in the comment?

simonqhughes replied:
If this work is restarted, this change can be made.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -65,21 +150,38 @@
+      PSA_CS_FILENAME_TID_PATTERN_LEN         /* TID (32-bit number in hex) */  \
      )
 
+/* PSA_CS_PATH_FILENAME_LENGTH is the buffer length for the full path and filename of a file
```
What is the length unit? Is it in bytes? Can you please specify it in the comment?

simonqhughes replied:
If this work is restarted, this change can be made.

simonqhughes committed:
SQ: Add <topdir>/test/psa-arch-tests patches and test/Makefile changes.

The following provides more information on this commit:
- The intend is to use the psa-arch-tests patches to patch the upstream repo
  to create the workspace. The developer fork will then not be required.
- The updating of the repo manifest to use upstream remains outstanding.

simonqhughes committed:
SQ: Various fixes in response to PR review comments.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -321,25 +490,38 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
+    ret = stat( src_filename, &file_stat );
+    if ( ret < 0 )
+    {
+        psa_debug( "Error: stat() failed with unexpected error (%d)\n", errno );
```
Error messages shouldn't be mixed with debug messages. If we don't compile with `PSA_STORAGE_DEBUG` we will miss important error messages.
I think you should have a `psa_error` or if you want a `psa_log` with log level feature (LOG_INFO, LOG_ERR etc).

simonqhughes replied:
This isn't an error message per-se, its a debug message. However, if this work is restarted, the debug log message can be changed to make this clear.

diego-sueiro commented on lib/common_storage.c:
```diff
@@ -29,29 +29,114 @@
 #include <string.h>
```
This source code file has ~4k lines of code.
I think it could be better modularized if you have the test code in a different file and the same thing for the recovery mechanism.

simonqhughes replied:
Agreed, this modularisation would be better. If this work is restarted, this work can be considered.

rwalton-arm commented on inc/config.h:
```diff
@@ -53,6 +53,10 @@
+ *
+ * Each application (process instance) has a maximum of PSA_STORAGE_FILE_MAX
+ * files. If N processes are active in the system then the maximum number
+ * of files is N * PSA_STORAGE_FILE_MAX.
```
How do we decide what value to give `PSA_STORAGE_FILE_MAX`? Is it arbitrary?

simonqhughes replied:
It depends on how many simultaneously open files you want in the system. Its not arbitrary but sized to the maximum number of storage objects that should ever be present in the system.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -262,19 +384,60 @@ static int psa_cs_dat_file_filter( const struct dirent *dir )
+    int len = strlen( s ) - cs_suffix_len;
+    const uid_t ruid = getuid();
+    const pid_t pid = getpid();
+    const pid_t tid = syscall(SYS_gettid);
```
We should check the return value for `syscall` and handle any errors explicitly. `syscall` will set `errno` and usually return -1 when it fails.

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -262,19 +384,60 @@ static int psa_cs_dat_file_filter( const struct dirent *dir )
+    const pid_t pid = getpid();
+    const pid_t tid = syscall(SYS_gettid);
+
+    snprintf( tmps_filter, PSA_CS_TMP_FILENAME_LENGTH, PSA_CS_TMP_FILENAME_RUID_PID_TID_PATTERN, (unsigned long int) ruid, (unsigned long int) pid, (unsigned long int) tid );
```
We should check the return value of `snprintf` to ensure the concatenation succeeded and the output wasn't truncated.

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -286,16 +449,18 @@ static psa_status_t psa_cs_get_mktemp_filename( char *filename, uint32_t len )
     psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+    const uid_t ruid = getuid();
+    const pid_t pid = getpid();
+    const pid_t tid = syscall(SYS_gettid);
```
We should check the return value for `syscall` and handle any errors explicitly. `syscall` will set `errno` and usually return -1 when it fails.

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -286,16 +449,18 @@ static psa_status_t psa_cs_get_mktemp_filename( char *filename, uint32_t len )
     dname = dirname( dup_fname );
-    snprintf( filename, len, "%s/%" PRIu64 "%s", dname, psa_cs_temp_file_counter, PSA_CS_BAD_FILE_SUFFIX );
-    psa_cs_temp_file_counter++;
+    snprintf( filename, len, "%s/" PSA_CS_TMP_FILENAME_RUID_PID_TID_PATTERN "%s", dname, (unsigned long int) ruid, (unsigned long int) pid, (unsigned long int) tid, PSA_CS_BAD_FILE_SUFFIX );
```
We should check the return value of `snprintf` to ensure the concatenation succeeded and the output wasn't truncated.

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -402,49 +589,301 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
-        ext = flags & PSA_CS_GET_FILENAME_F_DATA_FILE ? PSA_CS_DATA_FILE_SUFFIX: ext;
-        ext = flags & PSA_CS_GET_FILENAME_F_TEMP_FILE ? PSA_CS_TEMP_FILE_SUFFIX: ext;
-        snprintf( filename, PSA_CS_FILENAME_LENGTH,
+        snprintf( filename, len,
```
errors?

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -402,49 +589,301 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
+    else if( flags & PSA_CS_GET_FILENAME_F_TEMP_FILE )
+    {
+        pid = getpid();
+        tid = syscall(SYS_gettid);
```
errors?

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -402,49 +589,301 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
+    {
+        pid = getpid();
+        tid = syscall(SYS_gettid);
+        snprintf( filename, len,
```
errors?

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -402,49 +589,301 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
+    else if( flags & PSA_CS_GET_FILENAME_F_BAK_FILE )
     {
-        snprintf( filename, PSA_CS_FILENAME_LENGTH,
+        snprintf( filename, len,
```
errors?

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -402,49 +589,301 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
+    else if( flags & PSA_CS_GET_FILENAME_F_LOCK_FILE )
+    {
+        pid = getpid();
+        tid = syscall(SYS_gettid);
```
errors?

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -402,49 +589,301 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
+                          (unsigned long) ( uid >> 32 ),
+                          (unsigned long) ( uid & 0xffffffff ),
+                          PSA_CS_LOCK_FILE_SUFFIX );
+        status = PSA_SUCCESS;
```
Can we factor some of this repeated code out into helper functions?

simonqhughes replied:
Agreed. If this work is restarted, this work can be considered.

rwalton-arm commented on lib/common_storage.c:
```diff
@@ -402,49 +589,301 @@ static psa_status_t psa_cs_copy_file( const char *src_filename, const char *dst_
+    umask ( S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH );
+    if( fd_lock == NULL )
+    {
+        status = PSA_ERROR_INVALID_ARGUMENT;
```
should we log these errors?

simonqhughes replied:
Possibly. However, this would make the implementation tied to a logging system and less portable.

rwalton-arm reviewed (changes requested):
Haven't finished reviewing yet, but I have a few comments/questions.

simonqhughes commented:
This PR is being closed due to the closure of this feature development. See the relevant confluence pages for more information.
